### PR TITLE
[nan-339] protect against double inclusion of the base url

### PR DESCRIPTION
**Pull request by khaliqgant**
## Describe your changes
When paginating the base url gets included twice

## Issue ticket number and link
NAN-339

## Checklist before requesting a review (skip if just adding/editing APIs & templates)
- [ ] I added tests, otherwise the reason is: 
- [ ] I added observability, otherwise the reason is:
- [ ] I added analytics, otherwise the reason is:

### CODE REVIEW

File: packages/shared/lib/services/paginate.service.ts [hunk 1]
```diff
@@ -110,6 +112,11 @@ class PaginationService {
                 const url: URL = new URL(nextPageLink);
                 config.endpoint = url.pathname + url.search;
             }
+
+            const apiBase = config.baseUrlOverride || template.proxy?.base_url;
+            if (config.endpoint.includes(apiBase)) {
+                config.endpoint = config.endpoint.replace(apiBase, '');
+            }
```
Comment (TBonnin) [inline packages/shared/lib/services/paginate.service.ts]: not sure I follow. does that mean the returned nextPakelink contains twice the base url?

Comment (khaliqgant) [reply packages/shared/lib/services/paginate.service.ts]: That is correct

Comment (TBonnin) [reply packages/shared/lib/services/paginate.service.ts]: 😱 which provider is doing that?

**Review by TBonnin — APPROVED** [review]

**Review by bodinsamuel — APPROVED** [review]

File: packages/shared/lib/services/paginate.service.ts (hunk 1 above)
Comment (khaliqgant) [reply packages/shared/lib/services/paginate.service.ts]: Revisited this and this was actually a result of a regex that didn't catch it correctly. Added tests to protect against this.

File: packages/shared/lib/utils/utils.ts [hunk 2]
```diff
@@ -117,13 +117,14 @@ export function getRedisUrl() {
...
             '((\\d{1,3}\\.){3}\\d{1,3}))' + // OR ip (v4) address
-            '(\\:\\d+)?(\\/[-a-z\\d%_.~+]*)*' + // port and path
-            '(\\?[;&a-z\\d%_.~+=-]*)?' + // query string
-            '(\\#[-a-z\\d_]*)?$',
+            '(\\:\\d+)?(\\/[-a-z\\d%_.~+@]*)*' + // port and path
+            '(\\?[;&a-z\\d%_.~+=-@,]*)?' + // query string, including commas
+            '(\\#[-a-z\\d_]*)?$', // fragment locator
         'i'
-    ); // fragment locator
+    );
```
Comment (TBonnin) [inline packages/shared/lib/utils/utils.ts]: can we just use the `new URL()` instead of a regex?

File: packages/shared/lib/utils/utils.ts [hunk 3]
```diff
@@ -117,13 +117,14 @@ export function getRedisUrl() {
...
     return !!pattern.test(str);
 }
```
Comment (bodinsamuel) [inline packages/shared/lib/utils/utils.ts:124]: I think you could remove all of this in favor of native parsing, wdyt?

```js
try { 
  new URL(str);
  return true;
} catch {
  return false;
}
```

Comment (bodinsamuel) [reply packages/shared/lib/utils/utils.ts:124]: ah well @TBonnin was faster :D

File: packages/shared/lib/utils/utils.ts (hunk 2 above)
Comment (khaliqgant) [reply packages/shared/lib/utils/utils.ts]: Yeah, I suppose so. I'm unsure why the regex was used in the first place tbh but much easier to just use URL

File: packages/shared/lib/utils/utils.ts (hunk 3 above)
Comment (khaliqgant) [reply packages/shared/lib/utils/utils.ts:124]: haha, thanks, yeah, updated with https://github.com/NangoHQ/nango/pull/1638/commits/30851a46906bf3c376ffa38712453f42dc67e9ea

**Pull request merged**